pkg/renderer/sgml/renderer.go

```diff
@@ -0,0 +1,360 @@
+package sgml
+
+import (
+	"bytes"
+	htmltemplate "html/template"
+	"io"
+	"strings"
+	texttemplate "text/template"
+
+	"github.com/bytesparadise/libasciidoc/pkg/configuration"
+	"github.com/bytesparadise/libasciidoc/pkg/types"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// Renderer implements the backend render interface by using sgml.
+type Renderer interface {
+
+	// Render renders a document to the given output stream.
+	Render(ctx *Context, doc Document, output io.Writer) (Metadata, error)
+
+	// SetFunction sets the named function.
+	SetFunction(name string, fn interface{})
+
+	// Templates returns the Templates used by this Renderer.
+	// It cannot be altered on a given Renderer, since the old
+	// templates may have already been parsed.
+	Templates() Templates
+}
+
+func NewRenderer(t *Templates) Renderer {
+	sr := &sgmlRenderer{
+		functions: make(funcMap),
+		templates: t, // TODO: should we copy this?
+	}
+	// Establish some default function handlers.
+	sr.functions["render"] = sr.renderElements
+	sr.functions["renderElements"] = sr.renderElements
+	sr.functions["renderInline"] = sr.renderInlineElements
+	sr.functions["renderList"] = sr.renderListElements
+	sr.functions["renderLines"] = sr.renderLines
+	sr.functions["escape"] = EscapeString
+	sr.functions["renderToC"] = sr.renderTableOfContentsSections
+	sr.functions["renderFootnote"] = sr.renderFootnote
+	sr.functions["includeNewline"] = sr.includeNewline
+	sr.functions["renderVerse"] = sr.renderVerseBlockElement
+	sr.functions["plainText"] = sr.withPlainText
+	sr.functions["trimRight"] = sr.trimRight
+	sr.functions["trimLeft"] = sr.trimLeft
+	sr.functions["trim"] = sr.trimBoth
+	return sr
+}
+
+func (sr *sgmlRenderer) trimLeft(s string) string {
+	return strings.TrimLeft(s, " ")
+}
+
+func (sr *sgmlRenderer) trimRight(s string) string {
+	return strings.TrimRight(s, " ")
+}
+
+func (sr *sgmlRenderer) trimBoth(s string) string {
+	return strings.Trim(s, " ")
+}
+
+func (sr *sgmlRenderer) SetFunction(name string, fn interface{}) {
+	sr.functions[name] = fn
+}
+
+// Templates returns the Templates being used by this renderer.
+// A copy is made, as we cannot change the original Templates
+// due to it already being used.
+func (sr *sgmlRenderer) Templates() Templates {
+	return *sr.templates
+}
+
+func (sr *sgmlRenderer) newTemplate(name string, tmpl string, err error) (*textTemplate, error) {
+	// NB: if the data is missing below, it will be an empty string.
+	if err != nil {
+		return nil, err
+	}
+	t := texttemplate.New(name)
+	t.Funcs(sr.functions)
+	t, err = t.Parse(tmpl)
+	if err != nil {
+		log.Errorf("failed to initialize '%s' template: %v", name, err)
+		return nil, err
+	}
+	return t, nil
+}
+
+// Render renders the given document in HTML and writes the result in the given `writer`
+func (sr *sgmlRenderer) Render(ctx *Context, doc Document, output io.Writer) (Metadata, error) {
```

Comment:
shouldn't this be part of the `html5` package? 
same question for all subsequent funcs/methods in this file

Reply:
I don't think so.

The strategy here was to make html5 just a collection of templates, and then a thin wrapper of instantiation of an instance of the sgml.Renderer with the given templates.

The purpose of this is to make it very easy for people to build "custom" backends by just supplying their own templates.  Basically consider the templates as configuration (but very powerful configuration because they can contain code), so this separates the configuration from the code.  I think all SGML derived documents are probably going to be able to use this code more or less as is, with only modification to the templates -- although I believe html5s and DocBook might need us to enhance the parser to separate semantic differences so that we can for example render <SAMP> vs. <KBD> instead of just <PRE> for both.  (This will mean providing two different templates, but for HTML5 and XHTML they may well be the same template.)

To be more concrete, the XHTML and ePUB backends will be able to use the SGML, and will probably not need to modify any code except to provide the same thin wrappers.  (Actually ePUB will need to do a fair bit more, but that's because it has to build a document structure in a ZIP file, but it's going to use the SGML backend to build up the XHTML assets that are included in the epub.)

Of course, if a given backend (such as PDF) is going to need more differentiation here, then they can provide their own Render implementation.  That's going to be a lot more work, so I'm hoping the cases for it will be rare.

Reply:
yeah, that's what we discussed earlier indeed. My question was wrong, as I was looking very briefly at the changes here. I'll give it a deeper look later today ;)